### PR TITLE
Embedded lua scripts into FTL

### PR DESCRIPTION
**Pull request by DL6ER**
# What does this implement/fix?

Add facility to embed `lua` scripts into FTL. This reduces dependency on externally provided infrastructure and is easily extensible for to an arbitrary number of scripts whenever we deem this appropriate.

How to add new scripts? Just add them into the `src/lua/scripts/` directory and amend `CMakeLists.txt` as well as the code parts of FTL which are responsible for reading the scripts.

**Related issue or feature (if applicable):** N/A

**Pull request in [docs](https://github.com/pi-hole/docs) with documentation (if applicable):** N/A

---
**By submitting this pull request, I confirm the following:** 

1. I have read and understood the [contributors guide](https://docs.pi-hole.net/guides/github/contributing/), as well as this entire template. I understand which branch to base my commits and Pull Requests against. 
2. I have commented my proposed changes within the code.
3. I am willing to help maintain this change if there are issues with it later.
4. It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
5. I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))

## Checklist:

- [x] The code change is tested and works locally.
- [x] I based my code and PRs against the repositories `developmental` branch.
- [x] I [signed off](https://docs.pi-hole.net/guides/github/how-to-signoff/) all commits. Pi-hole enforces the [DCO](https://docs.pi-hole.net/guides/github/dco/) for all contributions
- [x] I [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) all my commits. Pi-hole requires signatures to verify authorship
- [x] I have read the above and my PR is ready for review.

### CODE REVIEW

File: src/lua/scripts/inspect.lua [hunk 1]
```diff
@@ -0,0 +1,335 @@
+local inspect ={
+  _VERSION = 'inspect.lua 3.1.0',
```
Comment (yubiuser) [inline src/lua/scripts/inspect.lua:2]: Maybe you also want to upgrade the embedded script?

https://github.com/kikito/inspect.lua/tags

**Review by yubiuser — COMMENTED** [review]
```
rockpi@rockpi-4b:~$ pihole-FTL -vv
****************************** FTL **********************************
Version:         vDev-395e079
Branch:          new/embedded_lua_scripts
Commit:          395e0792 (2022-11-20 13:53:41 +0100)
Architecture:    aarch64 (compiled on CI)
Compiler:        aarch64-linux-gnu-gcc (Debian 8.3.0-2) 8.3.0

****************************** dnsmasq ******************************
Version:         pi-hole-v2.88rc3
Compile options: IPv6 GNU-getopt no-DBus no-UBus no-i18n IDN DHCP DHCPv6 Lua TFTP no-conntrack ipset no-nftset auth cryptohash DNSSEC loop-detect inotify dumpfile

****************************** SQLite3 ******************************
Version:         3.40.0
Compile options: ATOMIC_INTRINSICS=1 COMPILER=gcc-8.3.0 DEFAULT_AUTOVACUUM DEFAULT_CACHE_SIZE=-16384 DEFAULT_FILE_FORMAT=4 DEFAULT_FOREIGN_KEYS DEFAULT_JOURNAL_SIZE_LIMIT=-1 DEFAULT_MEMSTATUS=0 DEFAULT_MMAP_SIZE=0 DEFAULT_PAGE_SIZE=4096 DEFAULT_PCACHE_INITSZ=20 DEFAULT_RECURSIVE_TRIGGERS DEFAULT_SECTOR_SIZE=4096 DEFAULT_SYNCHRONOUS=2 DEFAULT_WAL_AUTOCHECKPOINT=1000 DEFAULT_WAL_SYNCHRONOUS=2 DEFAULT_WORKER_THREADS=0 DQS=0 ENABLE_DBPAGE_VTAB MALLOC_SOFT_LIMIT=1024 MAX_ATTACHED=10 MAX_COLUMN=2000 MAX_COMPOUND_SELECT=500 MAX_DEFAULT_PAGE_SIZE=8192 MAX_EXPR_DEPTH=1000 MAX_FUNCTION_ARG=127 MAX_LENGTH=1000000000 MAX_LIKE_PATTERN_LENGTH=50000 MAX_MMAP_SIZE=0x7fff0000 MAX_PAGE_COUNT=1073741823 MAX_PAGE_SIZE=65536 MAX_SQL_LENGTH=1000000000 MAX_TRIGGER_DEPTH=1000 MAX_VARIABLE_NUMBER=32766 MAX_VDBE_OP=250000000 MAX_WORKER_THREADS=8 MUTEX_PTHREADS OMIT_DEPRECATED OMIT_DESERIALIZE OMIT_LOAD_EXTENSION OMIT_PROGRESS_CALLBACK SYSTEM_MALLOC TEMP_STORE=1 THREADSAFE=1
******************************** LUA ********************************
Lua 5.4.1  Copyright (C) 1994-2020 Lua.org, PUC-Rio

***************************** LIBNETTLE *****************************
Version: 3.8
GMP: Full
```

Just a thought: can we add which scripts are added (e.g. `inspect.lua`) to the `LUA` section in the version output?

File: src/lua/scripts/inspect.lua (hunk 1 above)
Comment (DL6ER) [reply src/lua/scripts/inspect.lua]: done

**Comment by DL6ER** [comment]
New `pihole-FTL -vv` output:
```
[...]

******************************** LUA ********************************
Lua 5.4.1  Copyright (C) 1994-2020 Lua.org, PUC-Rio
Embedded libraries: inspect.lua (8.46 KB) 

[...]
```

File: src/lua/scripts/inspect.lua [hunk 2]
```diff
@@ -0,0 +1,333 @@
+local _tl_compat; if (tonumber((_VERSION or ''):match('[%d.]*$')) or 0) < 5.3 then local p, m = pcall(require, 'compat53.module'); if p then _tl_compat = m end end; local math = _tl_compat and _tl_compat.math or math; local string = _tl_compat and _tl_compat.string or string; local table = _tl_compat and _tl_compat.table or table
```
Comment (rdwebdesign) [inline src/lua/scripts/inspect.lua:1]: Is this code in a single line intentionally?

Comment (DL6ER) [reply src/lua/scripts/inspect.lua:1]: Yes, see https://github.com/kikito/inspect.lua/blob/9c8a68da3033f1393b84b498313572654af82b8f/inspect.lua#L1

**Comment by yubiuser** [comment]
Should we add a test to check if all `*.lua` files in `src/lua/scripts` end up in the final binary?

**Comment by yubiuser** [comment]
This code can go now:

https://github.com/pi-hole/FTL/blob/844c8b9fda155b63e4bb8a43390bd4c23413d369/test/run.sh#L63-L65

File: src/lua/scripts/inspect.lua (hunk 2 above)
Comment (yubiuser) [reply src/lua/scripts/inspect.lua:1]: They also forgot to update their version string 

https://github.com/kikito/inspect.lua/blob/9c8a68da3033f1393b84b498313572654af82b8f/inspect.lua#L20

(should be `3.1.3` now)

**Comment by DL6ER** [comment]
@yubiuser This code is already removed in this PR:
Screenshot from 2022-11-25 19-22-27

I'm not sure a test will be necessary. What the user tries to do simply won't work if you don't include the new scripts. I'd rather like to ask for adding individual library tests very much like we have a test dedicated to test if the `inspect` library does what it should do. We can also very conveniently use this very same library to check the properties of any other future libraries we load and use by default.

File: src/lua/scripts/inspect.lua (hunk 2 above)
Comment (DL6ER) [reply src/lua/scripts/inspect.lua:1]: I'd vote for leaving it like it is. We have the most recent version included.

**Comment by yubiuser** [comment]
> @yubiuser This code is already removed in this PR: https://github.com/pi-hole/FTL/commit/6b6c069b227a45ff7f47ff8a91198d32eeb911e8

How did I miss this?

___

> I'd rather like to ask for adding individual library tests very much like we have a test dedicated to test if the inspect library does what it should do.

I'm fine with this approach, we only need to make sure we/others don't forget to add a test

**Review by yubiuser — APPROVED** [review]

**Comment by dschaper** [comment]
FTL already checks for `.lua` files in the lua path, correct? So this takes specific `.lua` files and bakes them in to the binary, no need for the file on the path.

What happens when there is a `.lua` file in the path that conflicts or duplicates what is in the binary? Say a version conflict.

**Pull request merged**